gdb/ppcfbsd-kern.c

```diff
@@ -75,10 +80,67 @@ ppcfbsd_supply_pcb(struct regcache *regcache, CORE_ADDR pcb_addr)
 			      (char *)&pcb.pcb_toc);
 
 	regcache->raw_supply(tdep->ppc_lr_regnum, (char *)&pcb.pcb_lr);
+	regcache->raw_supply(PPC_PC_REGNUM, (char *)&pcb.pcb_lr);
 	regcache->raw_supply(tdep->ppc_cr_regnum, (char *)&pcb.pcb_cr);
 }
 #endif
 
+static CORE_ADDR
+ppcfbsd_kern_disp(struct gdbarch *arch, const char *kernel, const char *vmcore)
+{
+	struct kvm_nlist nl[3];
+	struct gdbarch_tdep *tdep;
+	kvm_t *kd = NULL;
+	char errbuf[_POSIX2_LINE_MAX];
+	const char *errmsg = NULL;
+	CORE_ADDR kernbase;
+
+	if ((kd = kvm_openfiles(kernel, vmcore, NULL, O_RDONLY, errbuf)) ==
+	    NULL) {
+		errmsg = errbuf;
+		goto failed;
+	}
+
+	/*
+	 * Relocated KERNBASE should be read from __startkernel.
+	 * This works for core files, but live kernels fail to find it.
+	 * In this case, fallback to kernbase instead, that is resolved to
+	 * the relocated KERNBASE value on live kernels.
+	 * NOTE: resolving kernbase when using a core file results in
+	 *       KERNBASE value being returned instead.
+	 */
+	memset(nl, 0, sizeof(nl));
+	nl[0].n_name = "__startkernel";
+	nl[1].n_name = "kernbase";
+	nl[2].n_name = NULL;
+	errmsg = "Failed to locate kernel start symbol";
+	if (kvm_nlist2(kd, nl) == -1) {
+		goto failed;
+	} else if (nl[0].n_value != 0) {
+		tdep = gdbarch_tdep(arch);
+
+		if (kvm_read2(kd, nl[0].n_value, &kernbase, tdep->wordsize) ==
+		    -1) {
+			errmsg = "Failed to read kernel start symbol data";
+			goto failed;
+		}
+	} else if (nl[1].n_value != 0) {
+		kernbase = nl[1].n_value;
+	} else {
+		goto failed;
+	}
+
+	kvm_close(kd);
+	return kernbase - PPC_KERNBASE;
+
+failed:
+	warnx("warning: %s", errmsg);
```

Comment:
warnx() probably isn't the right thing to use in gdb as it has its own APIs like error() etc.

Reply:
Right, I'll replace warnx() with the equivalent gdb function.